File: ParallelAndVectorized/Algorithms/SynchronizedMatrixMultiplication.py
import numpy as np
import threading

class SynchronizedMatrixMultiplication:

    def multiply(self, A, B, threads):
        rows_A, cols_A = A.shape
        rows_B, cols_B = B.shape

        if cols_A != rows_B:
            raise ValueError("The number of columns in matrix A must be equal to the number of rows in matrix B.")

        result = np.zeros((rows_A, cols_B), dtype=np.float32)

        semaphore = threading.Semaphore(threads)

        row_counter = [0]

        def task():
            while True:
                with threading.Lock():
                    row = row_counter[0]
                    row_counter[0] += 1

                if row >= rows_A:
                    break

                for j in range(cols_B):
                    sum_ = 0
                    for k in range(cols_A):
                        sum_ += A[row, k] * B[k, j]
                    result[row, j] = sum_

                semaphore.release()

        threads_list = []
        for _ in range(threads):
            semaphore.acquire()
            t = threading.Thread(target=task)
            threads_list.append(t)
            t.start()

        for t in threads_list:
            t.join()

        return result

File: ParalellAndVectorizedMatrices/src/main/java/org/example/SynchronizedMatrixMultiplication.java
package org.example;

import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicInteger;

public class SynchronizedMatrixMultiplication implements ParallelMatrixMultiplication {

    @Override
    public double[][] multiply(double[][] A, double[][] B, int threads) throws InterruptedException {
        int rowsA = A.length;
        int colsA = A[0].length;
        int colsB = B[0].length;
        double[][] result = new double[rowsA][colsB];

        Semaphore semaphore = new Semaphore(threads);
        AtomicInteger rowCounter = new AtomicInteger(0);

        Runnable task = () -> {
            try {
                while (true) {
                    int row = rowCounter.getAndIncrement();
                    if (row >= rowsA) break;

                    for (int j = 0; j < colsB; j++) {
                        double sum = 0;
                        for (int k = 0; k < colsA; k++) {
                            sum += A[row][k] * B[k][j];
                        }
                        result[row][j] = sum;
                    }
                }
            } finally {
                semaphore.release();
            }
        };

        Thread[] threadPool = new Thread[threads];
        for (int i = 0; i < threads; i++) {
            semaphore.acquire();
            threadPool[i] = new Thread(task);
            threadPool[i].start();
        }

        for (Thread t : threadPool) {
            t.join();
        }

        return result;
    }
}
